homeassistant/components/mastodon/config_flow.py

```diff
@@ -0,0 +1,167 @@
+        vol.Required(
+            CONF_BASE_URL,
+            default=DEFAULT_URL,
+        ): TextSelector(TextSelectorConfig(type=TextSelectorType.URL)),
+        vol.Required(
+            CONF_CLIENT_ID,
+        ): TextSelector(TextSelectorConfig(type=TextSelectorType.PASSWORD)),
+        vol.Required(
+            CONF_CLIENT_SECRET,
+        ): TextSelector(TextSelectorConfig(type=TextSelectorType.PASSWORD)),
+        vol.Required(
+            CONF_ACCESS_TOKEN,
+        ): TextSelector(TextSelectorConfig(type=TextSelectorType.PASSWORD)),
+    }
+)
+
+
+class MastodonConfigFlow(ConfigFlow, domain=DOMAIN):
+    """Handle a config flow."""
+
+    VERSION = 1
+    config_entry: ConfigEntry
+
+    async def check_connection(
+        self,
+        base_url: str,
+        client_id: str,
+        client_secret: str,
+        access_token: str,
+    ) -> tuple[dict[str, str], dict[str, Any] | None]:
+        """Check connection to the Mastodon instance."""
+        try:
+            client = await self.hass.async_add_executor_job(
+                create_mastodon_instance,
+                base_url,
+                client_id,
+                client_secret,
+                access_token,
+            )
+            account = await self.hass.async_add_executor_job(
+                client.account_verify_credentials
+            )
+
+        except MastodonError:
+            return {"base": "credential_error"}, None
+        except Exception:  # noqa: BLE001
+            LOGGER.exception("Unexpected error")
+            return {"base": "unknown"}, None
+        return {}, account
+
+    def show_user_form(
+        self,
+        user_input: dict[str, Any] | None = None,
+        errors: dict[str, str] | None = None,
+        description_placeholders: dict[str, str] | None = None,
+        step_id: str = "user",
+    ) -> ConfigFlowResult:
+        """Show the user form."""
+        if user_input is None:
+            user_input = {}
+        return self.async_show_form(
+            step_id=step_id,
+            data_schema=self.add_suggested_values_to_schema(
+                STEP_USER_DATA_SCHEMA, user_input
+            ),
+            description_placeholders=description_placeholders,
+            errors=errors,
+        )
+
+    async def async_step_user(
+        self, user_input: dict[str, Any] | None = None
+    ) -> ConfigFlowResult:
+        """Handle a flow initialized by the user."""
+        if not user_input:
+            return self.show_user_form()
+
+        self._async_abort_entries_match({CONF_CLIENT_ID: user_input[CONF_CLIENT_ID]})
+
+        errors, account = await self.check_connection(
+            user_input[CONF_BASE_URL],
+            user_input[CONF_CLIENT_ID],
+            user_input[CONF_CLIENT_SECRET],
+            user_input[CONF_ACCESS_TOKEN],
+        )
+
+        if errors:
+            return self.show_user_form(user_input, errors)
+
+        await self.async_set_unique_id(user_input[CONF_CLIENT_ID])
```

Comment:
```suggestion
        await self.async_set_unique_id(user_input[CONF_CLIENT_ID])
        self._async_abort_unique_id_set()

        errors, account = await self.check_connection(
            user_input[CONF_BASE_URL],
            user_input[CONF_CLIENT_ID],
            user_input[CONF_CLIENT_SECRET],
            user_input[CONF_ACCESS_TOKEN],
        )

        if errors:
            return self.show_user_form(user_input, errors)
```
I think I still got the wrong function for the abort, but we can set the unique id and abort if needed

Reply:
Rewrote that whole method so hopefully it is clearer now.